modules/logwriter/variables.tf

```diff
@@ -0,0 +1,84 @@
+variable "name" {
+  type        = string
+  description = <<-EOF
+    Name of role. Since this name must be unique within the
+    account, it will be reused for most of the resources created by this
+    module.
+  EOF
+
+  validation {
+    condition     = length(var.name) <= 64
+    error_message = "Name must be under 64 characters."
+  }
+}
+variable "bucket_arn" {
+  description = "S3 Bucket ARN to write log records to."
+  type        = string
+  nullable    = false
+}
+
+variable "prefix" {
+  description = "Optional prefix to write log records to."
+  type        = string
+  nullable    = false
+  default     = "observe"
+}
+
+variable "log_group_name_patterns" {
+  description = "List of patterns as strings. We will only subscribe to log groups that have names matching one of the provided strings based on strings based on a case-sensitive substring search. To subscribe to all log groups, use the wildcard operator *."
+  type        = list(string)
+  nullable    = true
+  default     = [""]
+}
+
+variable "log_group_name_prefixes" {
+  description = "List of prefixes as strings. The lambda function will only apply to log groups that start with a provided string. To subscribe to all log groups, use the wildcard operator *."
+  type        = list(string)
+  nullable    = true
+  default     = [""]
+}
+
+variable "filter_name" {
+  description = "Subscription filter name. Existing filters that have this name as a prefix will be removed."
+  type        = string
+  nullable    = true
+  default     = "fake"
+}
+
+variable "filter_pattern" {
+  description = "Subscription filter pattern."
+  type        = string
+  nullable    = true
+  default     = ""
+}
+
+variable "buffering_interval" {
+  description = <<-EOF
+    Buffer incoming data for the specified period of time, in seconds, before
+    delivering it to S3.
+  EOF
+  type        = string
+  default     = 60
+}
+
+variable "buffering_size" {
+  description = <<-EOF
+    Buffer incoming data to the specified size, in MiBs, before delivering it
+    to S3.
+  EOF
+  type        = string
+  default     = 1
+}
+
+variable "num_workers" {
+  description = "Maximum number of concurrent workers when processing log groups."
+  type        = string
+  default     = 1
+}
+
+variable "discovery_rate" {
+  description = "EventBridge rate expression for periodically triggering discovery. If not set, no eventbridge rules are configured."
+  type        = string
+  nullable    = true
+  default     = ""
```

Comment:
You probably want to validate the string here to ensure it's a valid eventbridge rate expression.

Reply:
i found a regex online for this but its pretty gnarly. whats the downside to allowing the error from the aws provider to bubble up? in my testing that did happen

Reply:
I think this isn't so bad, because we're actually using EventBridge Scheduler rather than EventBridge Rules.

In Eventbridge rules, the [rate expression](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-rate-expressions.html) could be singular or plural (`1 minute` and `2 minutes` is fine, but never `2 minute` or `1 minutes`).

The fixed that madness with [EventBridge Scheduler.](https://docs.aws.amazon.com/scheduler/latest/UserGuide/schedule-types.html#rate-based). You now only use the pluralized form, so the regex would be `\d+ (minutes|hours|days)`.

The advantage of catching the error upfront is that it can get caught during `terraform plan`, so it's more friendly to CI. The disadvantage is that it may be too restrictive to future API changes (e.g. what if scheduler introduces `seconds`?).